Code review 1

alex-magana opened the pull request.

alex-magana committed:
Implement Song class

Implement the Song class. The Song class allows for creation of a class
and association of the class with an artist and a genre.

alex-magana committed:
Implement Artist class

Implement the Artist class. The Artist class allows for creation of a
class and association of the class with a song and genre.

alex-magana committed:
Implement Genre class

Implement the Genre class. The Genre class allows for creation of a class
and association of the class with a song and an artist.

alex-magana committed:
Implement the Concerns::Findable module

Implement the Concerns::Findable module. It exposes functionality to find
a song and create a song if it doesn't exist.

alex-magana committed:
Implement MusicImporter class

Implement MusicImporter class. It allows for creation of instances of Song
class when a file path to a directory of songs is passed.

alex-magana committed:
Implement MusicLibraryController class

Implement MusicLibraryController class. This class provides a CLI to allow
the user to interact with the directory of songs that is passed to the
class.

alex-magana committed:
Update Gemfile.lock

Include x86-mingw32 platform and upgrade bundle version.

alex-magana committed:
Refactor the MusicLibraryController class

Rewrite the class to expose a method for each CLI command. Implement
use of hash to evaluate user input and invoke the correct method.

alex-magana committed:
Rename the file

Rename the file to adhere to snake_case standard

alex-magana committed:
Rneame the file

Rename the file to adhere to snake_case.

alex-magana committed:
Include class description

Add a descriptioon of the functionality delivered by the class.

alex-magana committed:
Include class description

Add description indicating the functionality the class provides.

alex-magana committed:
Add class description

Include a description of the functionality provided by the class.

alex-magana committed:
Change require statement

Change require_relative to reflect the change in naming of song.rb

alex-magana committed:
Implement use of Concerns::Findable

Implement use of find_by_name method defined in Concerns::Findable

alex-magana committed:
Refactor the module

Refactor the module to enforce DRY.

alex-magana committed:
Refactor the code

Remodel the code to optimize methods.

alex-magana committed:
Invoke create_from_filename

Change the method used by import method.

alex-magana committed:
Refactor the code

Refactor the code to utilize the defined class relationships.

alex-magana committed:
Indicate addition and removal of a line

Indicate the addition and removal of a line

andela-mkimani commented on lib/Artist.rb:
```diff
@@ -0,0 +1,60 @@
+    end
+
+    def self.create(name)
+        Artist.new(name).save
```
Since you are inside the ```Artist``` class, you can simply call ```new``` instead of ```Artist.new```.

alex-magana committed:
Delete duplicate files

Remove MusicImporter.rb and MusicLibraryController.rb

andela-echigbo commented on lib/Song.rb:
```diff
@@ -0,0 +1,88 @@
+        song
+    end
+
+    def self.new_from_filename(file_name)
```
Any reason we are not using the `Findable` concern for this `find_or_create_by_name`?

ChigboIO commented on lib/Song.rb:
```diff
@@ -0,0 +1,88 @@
+        @@all.detect { |song| song.name == name}
+    end
+
+    # use Concerns::Findable to implement find_or_create_by_name
```
Any reason we are re-implementing this method that is already provided by the `Findable` concern?

alex-magana committed:
Remove unwanted \r

Remove unwanted invisible character \r in lib/Song.rb

ChigboIO commented on lib/Genre.rb:
```diff
@@ -0,0 +1,55 @@
+    end
+
+    def artists
+        @songs.map { |song| song.artist unless song.artist.nil? }.uniq
```
Given that, we defined an attribute accessor for the `@songs` attribute, I think we should `songs` here.

ChigboIO commented on lib/Artist.rb:
```diff
@@ -0,0 +1,60 @@
+    end
+
+    def genres
+       @songs.map { |song| song.genre unless song.genre.nil? }.uniq
```
Same comment as in  `Genre` class

ChigboIO commented on lib/Song.rb:
```diff
@@ -0,0 +1,89 @@
+        song = find_by_name(name)
+        song = Song.create(name) unless song
+        song
+    end
```
Any reason we are not using the `Findable` concern for this `find_or_create_by_name`?

ChigboIO commented on lib/music_library_controller.rb:
```diff
@@ -0,0 +1,99 @@
+        "\tlist genres\t: List all genres.\n" \
+        "\tplay song\t: Play a song.\n" \
+        "\tlist artist\t: List the artist.\n" \
+        "\tlist genre\t: List the genre.\n"
```
We should include the `exit` command to quit the program in this introduction

ChigboIO commented on lib/music_library_controller.rb:
```diff
@@ -0,0 +1,99 @@
+
+    attr_accessor   :path, :music_importer
+
+    @@commands = {"list songs" => :list_songs, 
```
You could make this `commands` a constant since it is just a constant hash of available commands

andela-echigbo commented:
Nice work man. Such a beautiful code you have here 😄 😄

morriskimani commented on lib/Song.rb:
```diff
@@ -0,0 +1,88 @@
+it to persist immediately.
+=end
+
+require_relative 'artist.rb'
```
Remove all the `require_relative`s. They have been handled for you in _config/environment.rb_

morriskimani commented on lib/Song.rb:
```diff
@@ -0,0 +1,88 @@

+    # extend Concerns::Findable

+
```
Why have you commented out `Concerns::Findable`?

morriskimani commented on lib/Song.rb:
```diff
@@ -0,0 +1,88 @@
+
+    attr_accessor   :name, :artist, :genre
+
+    @@all = []
```
For artist and genre, I suggest you use `attr_reader` since you have already defined getter methods `artist=` and `genre=`.

morriskimani commented on lib/Song.rb:
```diff
@@ -0,0 +1,88 @@
+    def self.create(name)
+        Song.new(name).save
+    end
+
```
You don't need to say `Song.new` since you are already in the `Song` class. `new` is just enough.

morriskimani commented on lib/Song.rb:
```diff
@@ -0,0 +1,88 @@
+
+    def artist=(_artist)
+        @artist = _artist
+        @artist.add_song(self)
```
Check on variables that begin with `_`

morriskimani commented on lib/Song.rb:
```diff
@@ -0,0 +1,88 @@
+    # use Concerns::Findable to implement find_by_name 
+    def self.find_by_name(name)
+        @@all.detect { |song| song.name == name}
+    end
```
`find_by_name` and `find_or_create_by_name` have already been defined in `Concerns::Findable` module. No need to redefine them.

morriskimani commented on lib/Song.rb:
```diff
@@ -0,0 +1,88 @@
+        #use Concerns::Findable to detect artist
+        artist = Artist.all.detect {|artist| artist.name == (song_details[0]).strip} || Artist.create((song_details[0]).strip)
+        #use Concerns::Findable to detect genre
+        genre = Genre.all.detect { |genre| genre.name == ((song_details[2]).gsub(".mp3", "")).strip } || Genre.create(((song_details[2]).gsub(".mp3", "")).strip)
```
This functionality has already been defined by the method `find_or_create_by_name` defined in module `Concerns::Findable` that has been extended to the `Artist` and `Genre` classes.

morriskimani commented on lib/Song.rb:
```diff
@@ -0,0 +1,88 @@
+=begin 
```
Is this comment really necessary?

morriskimani commented on lib/Artist.rb:
```diff
@@ -0,0 +1,60 @@
+it to persist immediately.
+=end
+
+require_relative '../concerns/concerns_findable.rb'
```
You don't have to require this file. This has already been handled in _config/environment.rb_.

morriskimani commented on lib/Artist.rb:
```diff
@@ -0,0 +1,60 @@
+=begin 
```
This comment is not necessary.

morriskimani commented on lib/Artist.rb:
```diff
@@ -0,0 +1,60 @@
+        @songs = []
+    end
+
+    def self.create(name)
```
You can simply use `new(name).save` since you are already in the `Artist` class.

morriskimani commented on lib/Genre.rb:
```diff
@@ -0,0 +1,55 @@
+it to persist immediately.
+=end
+
+require_relative '../concerns/concerns_findable.rb'
```
You don't have to require this file. This has already been handled in _config/environment.rb_.

morriskimani commented on lib/music_importer.rb:
```diff
@@ -0,0 +1,22 @@
+require_relative 'song.rb'
```
You dont need to require.

morriskimani commented on lib/music_importer.rb:
```diff
@@ -0,0 +1,22 @@
+
+    def initialize(path)
+        @path = path
+        @file_names = []
```
No need for this array since we are dynamically retrieving filenames using the `files` method defined below.

morriskimani commented on lib/music_importer.rb:
```diff
@@ -0,0 +1,22 @@
+
+    def import
+        files_all = files
+        files_all.each { |file| Song.create_from_filename(file) }
```
Why not `files.each` ... ? This way you don't even need the variable `files_all`.

morriskimani commented on lib/music_library_controller.rb:
```diff
@@ -0,0 +1,99 @@
+input. Check out the tests for specifics
+=end
+
+require_relative 'music_importer.rb'
```
You don't have to require this file.

morriskimani commented on lib/music_library_controller.rb:
```diff
@@ -0,0 +1,99 @@
+        
+        while true
+            print PROMPT
+            user_input = self.send(:gets).chomp
```
Why not simply say `gets.chomp`?

morriskimani replied:
What if the user input has trailing whitespaces? Try stripping the `user_input`.

morriskimani commented on lib/music_library_controller.rb:
```diff
@@ -0,0 +1,99 @@
+        self.command_execute(@@commands[command_name])
+    end
+
+    def command_execute(command_name)
```
You have many unnecessary methods. You don't really need the functions `command_evaluate` and `command_execute`.

morriskimani commented on lib/music_library_controller.rb:
```diff
@@ -0,0 +1,99 @@
+    def play_song
+        puts "Enter the number of the song to play."
+        song_number = self.send(:gets).chomp.to_i
+        song = Song.all[song_number - 1]
```
What if a user enters a number that is beyond the array  range or enters a string? Try to handle invalid input.

alex-magana committed:
Add Travis CI integration

Add Travis CI yml file to the repo.

alex-magana committed:
Remove Travis Ci .travis.yml file

alex-magana committed:
Add Travis CI integration

Add Travis CI .travis.yml file to the repository.

alex-magana committed:
Change file names and move concerns directory

Rename lib/Artist.rb to lib/artist.rb
Rename lib/Genre.rb to lib/genre.rb
Rename lib/Song.rb to lib/song.rb
Move concerns directory to lib/concerns/

alex-magana committed:
Refactor class MusicImporter

Eliminate file_names array. Eliminate repetition in method files.
Return array on files method call without variable declaration.

alex-magana committed:
Refactor class MusicLibraryController

Make the commands hash a constant. Join the command_evaluate and
command_execute.

alex-magana committed:
Refactor class Song

Extend Concerns::Findable.
Declare attr_reader for artist and genre instance variables.
Eliminate methods find_by_name and find_or_create_by_name.
Remove unnecessary repetition in method new_from_filename.

alex-magana committed:
Rectify code to adhere to standards

Use 2 spaces to organize code. Check file names adhere to the
naming format that uses - (hypen).

alex-magana committed:
Correct the indentation

Use 2 spaces instead of 4 spaces.

alex-magana committed:
Correct the indentation

Use 2 spaces instead of 4 spaces.

alex-magana committed:
Correct the indentation

Use 2 spaces instead of 4 spaces.

alex-magana committed:
Correct indentation and filter path files

Use 2 spaces instead of 4 spaces. Use method glob to return .mp3
files from the path provided.

Combine command_avaluate and command_execute. Include help method
to suport wrong user input. Implement check for song number validity
for command play song.

alex-magana committed:
Add coveralls.

Implement coveralls to monitor the scope covered by tests.

alex-magana committed:
Implement module for shared class methods.

alex-magana committed:
Implement module for shared nstance methods.

alex-magana committed:
Add view class to complete MVC implementation.

alex-magana committed:
Add implementation for coveralls and colorize

alex-magana committed:
Use modules to implement shared methods

alex-magana committed:
Organize code using 2 spaces in place of tabs

alex-magana committed:
Remove unused variables

alex-magana committed:
Modify controller to invoke view methods

alex-magana committed:
Commit the last test coverage check

alex-magana committed:
Change the application name

Change the application name as seen in the introduction text.

alex-magana committed:
Update README.md

alex-magana committed:
Update README.md

alex-magana committed:
Update README.md

alex-magana committed:
Update README.md

alex-magana committed:
Update README.md

alex-magana committed:
Update README.md

alex-magana committed:
Update README.md

alex-magana committed:
Update README.md
